<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Home · BinBencher.jl</title><meta name="title" content="Home · BinBencher.jl"/><meta property="og:title" content="Home · BinBencher.jl"/><meta property="twitter:title" content="Home · BinBencher.jl"/><meta name="description" content="Documentation for BinBencher.jl."/><meta property="og:description" content="Documentation for BinBencher.jl."/><meta property="twitter:description" content="Documentation for BinBencher.jl."/><script data-outdated-warner src="assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="search_index.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href>BinBencher.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li class="is-active"><a class="tocitem" href>Home</a><ul class="internal"><li><a class="tocitem" href="#Nomenclariture-of-BinBencher"><span>Nomenclariture of BinBencher</span></a></li><li><a class="tocitem" href="#Derived-metrics"><span>Derived metrics</span></a></li><li><a class="tocitem" href="#input_files"><span>Input file formats</span></a></li></ul></li><li><span class="tocitem">CLI interface</span><ul><li><a class="tocitem" href="cli_overview/">Overview</a></li><li><a class="tocitem" href="benchmarking/">Benchmarking</a></li><li><a class="tocitem" href="make_ref/">Making a reference</a></li></ul></li><li><span class="tocitem">Backend package</span><ul><li><a class="tocitem" href="backend_overview/">Overview</a></li><li><a class="tocitem" href="walkthrough/">Walkthrough</a></li><li><a class="tocitem" href="reference/">API Reference</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Home</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Home</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/jakobnissen/BinBencherBackend.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/jakobnissen/BinBencherBackend.jl/blob/master/docs/src/index.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="BinBencher"><a class="docs-heading-anchor" href="#BinBencher">BinBencher</a><a id="BinBencher-1"></a><a class="docs-heading-anchor-permalink" href="#BinBencher" title="Permalink"></a></h1><p>BinBencher is a tool for analysis and benchmarking of a set of metagenomic assembled genomes (MAGs) against a reference. It is designed to be used for benchmarking metagenomic binners against a metagenome with a known ground truth, typically from a simulated metagenome such as the ones from <a href="https://github.com/CAMI-challenge/CAMISIM">CAMISIM</a>, or alternatively from a laboratory-created mock community. </p><p>BinBencher is <a href="https://www.biorxiv.org/content/10.1101/2024.05.06.592671v1.full.pdf">pre-published at bioRxiv</a>. Read that paper for background and motivation for this package.</p><p>BinBencher consists of two packages:</p><ul><li><code>BinBencherBackend.jl</code>: The Julia library that does the actual analysis.</li><li><code>BinBencher.jl</code>: An application that wraps the backend in a command-line interface.</li></ul><p>Most users probably want to use the CLI wrapper.</p><h2 id="Nomenclariture-of-BinBencher"><a class="docs-heading-anchor" href="#Nomenclariture-of-BinBencher">Nomenclariture of BinBencher</a><a id="Nomenclariture-of-BinBencher-1"></a><a class="docs-heading-anchor-permalink" href="#Nomenclariture-of-BinBencher" title="Permalink"></a></h2><h3 id="Basic-nomenclariture"><a class="docs-heading-anchor" href="#Basic-nomenclariture">Basic nomenclariture</a><a id="Basic-nomenclariture-1"></a><a class="docs-heading-anchor-permalink" href="#Basic-nomenclariture" title="Permalink"></a></h3><ul><li>A <strong>sequence</strong> is the sequences that are to be clustered by the binner. I also sometimes call this a <strong>contig</strong>, because in most applications, binners are used to bin contigs.</li><li>A <strong>genome</strong> is a target genome that should be reconstructed by the binner. It can be a virus, organism, plasmid or any other genetic entity. Every genome have one or more sources, zero or more flags, and one parent clade.</li><li>A <strong>flag</strong> marks the certaincy about a boolean attribute of a genome, like &quot;is this a virus?&quot;.</li><li>A <strong>source</strong> are the DNA sequences that genomes are composed of. E.g. for a bacteria, its sources are all its chromosomes. For plasmids, its only source is its full-length sequence.  Sources are typically the full genome reference sequences from e.g. NCBI&#39;s RefSeq. Sequences map to zero or more sources at particular spans, i.e. a contiguous range from a starting to a stopping position.</li><li>A <strong>clade</strong> contain one or more genomes, or clades, known as its children. Clades containing genomes are rank 1, and clades containing rank N clades are rank N+1 clades. By extension, genomes have rank 0. A clade never has children of mixed ranks. All genomes descend from a chain of exactly N ranks of clades, where N &gt; 0.</li><li>A <strong>bin</strong> is a set of sequences created by the binner. Every bin is benchmarked against all genomes and clades in the reference.</li><li>A <strong>reference</strong> is composed of:<ul><li>The set of genomes__ to benchmark against, which all ultimately descend from a single common clades that links all genomes together.</li><li>Its set of sequences each with zero or more mappings to its genomes&#39; sources.</li></ul></li><li>A <strong>binning</strong> is a set of bins benchmarked against a reference, where every bin is a subset of the reference&#39;s sequences.</li></ul><p>In the equations below, we denote genomes <span>$G$</span>. Genomes can be considered disjoint sets of mapping positions (i.e. genomic positions), where each position is a basepair in a source of the genome. The total set of positions in the reference is <span>$Y = \cup_{G}$</span>. Let <span>$X$</span> be the set of sequences <span>$S$</span> to be binned. Each sequence has a length <span>$L_S$</span> and can be considered as a set of mapping positions <span>$S \subseteq Y$</span>. Note that the cardinality of this set <span>$|S|$</span> need not be equal to the sequence&#39;s length <span>$L_S$</span>.</p><p>If we have a set of sequences <span>$x \subseteq X$</span> and a set of mapping positions <span>$y \subseteq Y$</span>, let us define <span>$x \doublecap y := \{S \in x | S \cap y \neq \emptyset \}$</span>, i.e. the set of sequences in <span>$x$</span> that have mapping positions in <span>$y$</span>.</p><p>A genome&#39;s <em>assembly</em> <span>$A_G = G \cap \cup_X$</span> are all mapping positions in the genome covered by any sequence in the dataset.</p><p>A bin <span>$B$</span> is a set of sequences <span>$B \subseteq X$</span>. For any bin-genome pair <span>$\{B,G\}$</span>, we define:</p><ul><li>The true positives <span>$TP_{\{B,G\}} = |\cup_{S \in B} S \cap G|$</span>, the number of mapping positions in <span>$G$</span> that any sequence of <span>$B$</span> maps to.</li><li>The false positives <span>$FP_{\{B,G\}} = \sum_{S \in G \setminus (B \doublecap G)} L_S$</span> is the sum of lengths of sequences in <span>$B$</span> that does not map to any position in <span>$G$</span>.</li><li>The false <em>genomic</em> negatives <span>$FG_{\{B,G\}} = |G| - TP_{\{B,G\}}$</span> is all positions in <span>$G$</span> not covered by any sequence in <span>$B$</span>.</li><li>The false <em>assembly</em> negatives <span>$FA_{\{B,G\}} = |A_G| - TP_{\{B,G\}}$</span> is the assembly of G not covered by any sequence in <span>$B$</span>.</li></ul><h2 id="Derived-metrics"><a class="docs-heading-anchor" href="#Derived-metrics">Derived metrics</a><a id="Derived-metrics-1"></a><a class="docs-heading-anchor-permalink" href="#Derived-metrics" title="Permalink"></a></h2><p>BinBencher computes a bunch of different metrics. The most central ones are given here. More may be added in the future.</p><ul><li>The precision of a bin-genome pair is <span>$\frac{TP_{\{B,G\}}}{TP_{\{B,G\}} + FP_{\{B,G\}}}$</span></li><li>The <em>genomic recall</em> is <span>$\frac{TP_{\{B,G\}}}{TP_{\{B,G\}} + FG_{\{B,G\}}}$</span></li><li>The <em>assembly recall</em> is <span>$\frac{TP_{\{B,G\}}}{TP_{\{B,G\}} + FA_{\{B,G\}}}$</span></li></ul><p>Note that</p><ul><li>These metrics are only defined for bin-genome pairs. It is not sensical to talk about the precision of a genome without specifying with respect to which bin.</li><li>The assembly recall is never lower than the genomic recall.</li></ul><p>The most useful measures are the number of recovered bins or genomes at a given recall/precision threshold. This counts the total number of genomes (bins, respectively) which, when paired with any bin (genomes, resp.) has a recall and precision over the threhold. Both these measures can be computed for assembly or genomic recall.</p><h2 id="input_files"><a class="docs-heading-anchor" href="#input_files">Input file formats</a><a id="input_files-1"></a><a class="docs-heading-anchor-permalink" href="#input_files" title="Permalink"></a></h2><h3 id="Binning-TSV-file"><a class="docs-heading-anchor" href="#Binning-TSV-file">Binning TSV file</a><a id="Binning-TSV-file-1"></a><a class="docs-heading-anchor-permalink" href="#Binning-TSV-file" title="Permalink"></a></h3><p>The binning is a tab-separated values (TSV) file of the following format:</p><ul><li>The header must be exactly <code>clustername\tcontigname</code></li><li>Every subsequent line must be the clustername, then a tab, then the contigname.</li></ul><p>The format can be written as a regular expression:</p><pre><code class="nohighlight hljs">binning = &quot;clustername\tcontigname&quot; (&quot;\r?\n&quot; bin)* (&quot;\r?\n&quot;)?
bin = identifier &quot;\t&quot; identifier
identifier = &quot;[^\t\r\n]*&quot;</code></pre><p>The file is assumed to be encoded in UTF-8. The identifiers may contain any arbitrary bytes except &#39;\t&#39; (0x09), &#39;\n&#39; (0x0a) or &#39;\r&#39; (0x0d), including invalid unicode.</p><h3 id="refjson"><a class="docs-heading-anchor" href="#refjson">Reference JSON file</a><a id="refjson-1"></a><a class="docs-heading-anchor-permalink" href="#refjson" title="Permalink"></a></h3><p>This reference file is complex, as it contains a wealth of information about the genomes and their relationship. It is recommended to create this file using the CLI - see the relevant section of this documentation.</p><p>If you want to create the JSON file yourself, you can learn its format by reading files in <a href="https://github.com/jakobnissen/BinBencherBackend.jl">the BinBencherBackend.jl repository</a>:</p><ul><li>A JSON Schema with descriptions at <code>files/schema.json</code></li><li>A small example reference in <code>files/ref.json</code></li><li>The source code in <code>src/reference.json</code> - grep for <code>ReferenceJSON</code></li></ul></article><nav class="docs-footer"><a class="docs-footer-nextpage" href="cli_overview/">Overview »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.6.0 on <span class="colophon-date" title="Tuesday 3 September 2024 12:00">Tuesday 3 September 2024</span>. Using Julia version 1.10.5.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
